File: binary-indexed-tree/BinaryIndexedTreeTest.cpp
/* 
 * Binary indexed tree test (C++)
 * 
 * Copyright (c) 2022 Project Nayuki. (MIT License)
 * https://www.nayuki.io/page/binary-indexed-tree
 * 
 * Permission is hereby granted, free of charge, to any person obtaining a copy of
 * this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to
 * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
 * the Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 * - The above copyright notice and this permission notice shall be included in
 *   all copies or substantial portions of the Software.
 * - The Software is provided "as is", without warranty of any kind, express or
 *   implied, including but not limited to the warranties of merchantability,
 *   fitness for a particular purpose and noninfringement. In no event shall the
 *   authors or copyright holders be liable for any claim, damages or other
 *   liability, whether in an action of contract, tort or otherwise, arising from,
 *   out of or in connection with the Software or the use or other dealings in the
 *   Software.
 */

#include <algorithm>
#include <cstddef>
#include <cstdint>
#include <cstdlib>
#include <exception>
#include <iostream>
#include <limits>
#include <random>
#include <vector>
#include "BinaryIndexedTree.hpp"

using std::size_t;
using std::vector;
using std::uniform_int_distribution;


// Forward declarations
static void testSizeConstructor();
static void testAllOnes();
static void testArrayConstructorRandomly();
static void testAddAndSetRandomly();


// Random number generation global variables
std::default_random_engine randGen((std::random_device())());


int main() {
	try {
		testSizeConstructor();
		testAllOnes();
		testArrayConstructorRandomly();
		testAddAndSetRandomly();
		
		std::cerr << "Test passed" << std::endl;
		return EXIT_SUCCESS;
	} catch (std::exception &e) {
		std::cerr << e.what() << std::endl;
		return EXIT_FAILURE;
	}
}


template <typename T>
static void assertEquals(T x, T y) {
	if (x != y)
		throw std::runtime_error("Value mismatch");
}


static void testSizeConstructor() {
	const size_t SIZELIMIT = 10000;
	const long CHECKS = 10;
	using T = std::int8_t;
	for (size_t len = 0; len < SIZELIMIT; len++) {
		
		const BinaryIndexedTree<T> bt(len);
		assertEquals(len, bt.size());
		assertEquals(static_cast<T>(0), bt.getTotal());
		
		uniform_int_distribution<size_t> indexDist(0, len > 0 ? len - 1 : 0);
		uniform_int_distribution<size_t> indexOneDist(0, len);
		for (long i = 0; i < CHECKS; i++) {
			if (len > 0)
				assertEquals(static_cast<T>(0), bt[indexDist(randGen)]);
			assertEquals(static_cast<T>(0), bt.getPrefixSum(indexOneDist(randGen)));
			
			size_t start = indexOneDist(randGen);
			size_t end   = indexOneDist(randGen);
			if (start > end)
				std::swap(start, end);
			assertEquals(static_cast<T>(0), bt.getRangeSum(start, end));
		}
	}
}


static void testAllOnes() {
	const size_t SIZELIMIT = 10000;
	const long CHECKS = 10;
	using T = std::uint16_t;
	uniform_int_distribution<int> modeDist(0, 3);
	std::uniform_real_distribution<double> realDist;
	for (size_t len = 1; len < SIZELIMIT; len++) {
		
		BinaryIndexedTree<T> bt(0);
		int mode = modeDist(randGen);
		if (mode == 0) {
			vector<T> vals(len, 1);
			bt = BinaryIndexedTree<T>(vals);
		} else {
			bt = BinaryIndexedTree<T>(len);
			double p;
			if      (mode == 1) p = 0;
			else if (mode == 2) p = 1;
			else if (mode == 3) p = realDist(randGen);
			else throw std::domain_error("Assertion error");
			for (size_t i = 0; i < len; i++) {
				if (realDist(randGen) < p)
					bt.add(i, 1);
				else
					bt.set(i, 1);
			}
		}
		
		assertEquals(len, bt.size());
		assertEquals(static_cast<T>(len), bt.getTotal());
		uniform_int_distribution<size_t> indexDist(0, len - 1);
		uniform_int_distribution<size_t> indexOneDist(0, len);
		for (long i = 0; i < CHECKS; i++) {
			assertEquals(static_cast<T>(1), bt[indexDist(randGen)]);
			size_t k = indexOneDist(randGen);
			assertEquals(static_cast<T>(k), bt.getPrefixSum(k));
			
			size_t start = indexOneDist(randGen);
			size_t end   = indexOneDist(randGen);
			if (start > end)
				std::swap(start, end);
			assertEquals(static_cast<T>(end - start), bt.getRangeSum(start, end));
		}
	}
}


static void testArrayConstructorRandomly() {
	const long TRIALS = 10000;
	const size_t SIZELIMIT = 10000;
	const long CHECKS = 100;
	using T = std::int64_t;
	uniform_int_distribution<size_t> lenDist(0, SIZELIMIT);
	uniform_int_distribution<T> valDist(INT64_C(-1000000), INT64_C(1000000));
	for (long i = 0; i < TRIALS; i++) {
		
		size_t len = lenDist(randGen);
		vector<T> vals;
		vector<T> cums{0};
		for (size_t j = 0; j < len; j++) {
			vals.push_back(valDist(randGen));
			cums.push_back(cums.back() + vals.back());
		}
		
		BinaryIndexedTree<T> bt(vals);
		assertEquals(len, bt.size());
		assertEquals(cums.at(len), bt.getTotal());
		
		uniform_int_distribution<size_t> indexDist(0, len > 0 ? len - 1 : 0);
		uniform_int_distribution<size_t> indexOneDist(0, len);
		for (long j = 0; j < CHECKS; j++) {
			if (len > 0) {
				size_t k = indexDist(randGen);
				assertEquals(vals.at(k), bt[k]);
			}
			size_t k = indexOneDist(randGen);
			assertEquals(cums.at(k), bt.getPrefixSum(k));
			
			size_t start = indexOneDist(randGen);
			size_t end   = indexOneDist(randGen);
			if (start > end)
				std::swap(start, end);
			assertEquals(cums.at(end) - cums.at(start), bt.getRangeSum(start, end));
		}
	}
}


static void testAddAndSetRandomly() {
	const long TRIALS = 10000;
	const long SIZELIMIT = 10000;
	const long OPERATIONS = 10000;
	const long CHECKS = 100;
	uniform_int_distribution<size_t> lenDist(1, SIZELIMIT);
	std::bernoulli_distribution modeDist;
	using T = std::uint64_t;
	uniform_int_distribution<T> valDist(0, std::numeric_limits<T>::max());
	for (long i = 0; i < TRIALS; i++) {
		
		size_t len = lenDist(randGen);
		vector<T> vals;
		BinaryIndexedTree<T> bt(0);
		if (modeDist(randGen)) {
			vals = vector<T>(len, 0);
			bt = BinaryIndexedTree<T>(len);
		} else {
			for (size_t j = 0; j < len; j++)
				vals.push_back(valDist(randGen));
			bt = BinaryIndexedTree<T>(vals);
		}
		
		uniform_int_distribution<size_t> indexDist(0, len > 0 ? len - 1 : 0);
		for (long j = 0; j < OPERATIONS; j++) {
			size_t k = indexDist(randGen);
			T x = valDist(randGen);
			if (modeDist(randGen)) {
				vals.at(k) += x;
				bt.add(k, x);
			} else {
				vals.at(k) = x;
				bt.set(k, x);
			}
		}
		
		vector<T> cums{0};
		for (T x : vals)
			cums.push_back(cums.back() + x);
		
		uniform_int_distribution<size_t> indexOneDist(0, len);
		for (long j = 0; j < CHECKS; j++) {
			size_t k = indexDist(randGen);
			assertEquals(vals.at(k), bt[k]);
			k = indexOneDist(randGen);
			assertEquals(cums.at(k), bt.getPrefixSum(k));
			
			size_t start = indexOneDist(randGen);
			size_t end   = indexOneDist(randGen);
			if (start > end)
				std::swap(start, end);
			assertEquals(cums.at(end) - cums.at(start), bt.getRangeSum(start, end));
		}
	}
}

File: binary-indexed-tree/BinaryIndexedTreeTest.ts
/* 
 * Binary indexed tree test (TypeScript)
 * 
 * Copyright (c) 2022 Project Nayuki. (MIT License)
 * https://www.nayuki.io/page/binary-indexed-tree
 * 
 * Permission is hereby granted, free of charge, to any person obtaining a copy of
 * this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to
 * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
 * the Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 * - The above copyright notice and this permission notice shall be included in
 *   all copies or substantial portions of the Software.
 * - The Software is provided "as is", without warranty of any kind, express or
 *   implied, including but not limited to the warranties of merchantability,
 *   fitness for a particular purpose and noninfringement. In no event shall the
 *   authors or copyright holders be liable for any claim, damages or other
 *   liability, whether in an action of contract, tort or otherwise, arising from,
 *   out of or in connection with the Software or the use or other dealings in the
 *   Software.
 */


/*---- Test suite ----*/

const TEST_SUITE_FUNCS: Array<()=>void> = [
	
	function testSizeConstructor(): void {
		const SIZELIMIT: number = 3000;
		const CHECKS: number = 10;
		for (let len = 0; len < SIZELIMIT; len++) {
			
			let bt = new BinaryIndexedTree(len);
			assertEquals(len, bt.length);
			assertEquals(0, bt.getTotal());
			
			for (let i = 0; i < CHECKS; i++) {
				if (len > 0)
					assertEquals(0, bt.get(randInt(len)));
				assertEquals(0, bt.getPrefixSum(randInt(len + 1)));
				
				let start: number = randInt(len + 1);
				let end  : number = randInt(len + 1);
				if (start > end) {
					const temp: number = start;
					start = end;
					end = temp;
				}
				assertEquals(0, bt.getRangeSum(start, end));
			}
		}
	},
	
	
	function testAllOnes(): void {
		const SIZELIMIT: number = 3000;
		const CHECKS: number = 10;
		for (let len = 1; len < SIZELIMIT; len++) {
			
			let bt: BinaryIndexedTree;
			const mode: number = randInt(4);
			if (mode == 0) {
				let vals: Array<number> = [];
				for (let i = 0; i < len; i++)
					vals.push(1);
				bt = new BinaryIndexedTree(vals);
			} else {
				bt = new BinaryIndexedTree(len);
				let p: number;
				if      (mode == 1) p = 0;
				else if (mode == 2) p = 1;
				else if (mode == 3) p = Math.random();
				else throw new Error("Assertion error");
				for (let i = 0; i < len; i++) {
					if (Math.random() < p)
						bt.add(i, 1);
					else
						bt.set(i, 1);
				}
			}
			
			assertEquals(len, bt.length);
			assertEquals(len, bt.getTotal());
			for (let i = 0; i < CHECKS; i++) {
				assertEquals(1, bt.get(randInt(len)));
				const k: number = randInt(len + 1);
				assertEquals(k, bt.getPrefixSum(k));
				
				let start: number = randInt(len + 1);
				let end  : number = randInt(len + 1);
				if (start > end) {
					const temp: number = start;
					start = end;
					end = temp;
				}
				assertEquals(end - start, bt.getRangeSum(start, end));
			}
		}
	},
	
	
	function testArrayConstructorRandomly(): void {
		const TRIALS: number = 3000;
		const SIZELIMIT: number = 10000;
		const CHECKS: number = 100;
		for (let i = 0; i < TRIALS; i++) {
			
			const len: number = randInt(SIZELIMIT);
			let vals: Array<number> = [];
			let cums: Array<number> = [0];
			for (let j = 0; j < len; j++) {
				vals.push(randInt(2001) - 1000);
				cums.push(cums[j] + vals[j]);
			}
			
			let bt = new BinaryIndexedTree(vals);
			assertEquals(len, bt.length);
			assertEquals(cums[len], bt.getTotal());
			
			for (let j = 0; j < CHECKS; j++) {
				if (len > 0) {
					const k: number = randInt(len);
					assertEquals(vals[k], bt.get(k));
				}
				const k: number = randInt(len + 1);
				assertEquals(cums[k], bt.getPrefixSum(k));
				
				let start: number = randInt(len + 1);
				let end  : number = randInt(len + 1);
				if (start > end) {
					const temp: number = start;
					start = end;
					end = temp;
				}
				assertEquals(cums[end] - cums[start], bt.getRangeSum(start, end));
			}
		}
	},
	
	
	function testAddAndSetRandomly(): void {
		const TRIALS: number = 3000;
		const SIZELIMIT: number = 3000;
		const OPERATIONS: number = 3000;
		const CHECKS: number = 100;
		for (let i = 0; i < TRIALS; i++) {
			
			const len: number = randInt(SIZELIMIT) + 1;
			let vals: Array<number> = [];
			let bt: BinaryIndexedTree;
			if (Math.random() < 0.5) {
				for (let j = 0; j < len; j++)
					vals.push(0);
				bt = new BinaryIndexedTree(len);
			} else {
				for (let j = 0; j < len; j++)
					vals[j] = randInt(2001) - 1000;
				bt = new BinaryIndexedTree(vals);
			}
			
			for (let j = 0; j < OPERATIONS; j++) {
				const k: number = randInt(len);
				const x: number = randInt(2001) - 1000;
				if (Math.random() < 0.5) {
					vals[k] += x;
					bt.add(k, x);
				} else {
					vals[k] = x;
					bt.set(k, x);
				}
			}
			
			let cums: Array<number> = [0];
			vals.forEach((x, j) =>
				cums.push(cums[j] + x));
			
			for (let j = 0; j < CHECKS; j++) {
				let k: number = randInt(len);
				assertEquals(vals[k], bt.get(k));
				k = randInt(len + 1);
				assertEquals(cums[k], bt.getPrefixSum(k));
				
				let start: number = randInt(len + 1);
				let end  : number = randInt(len + 1);
				if (start > end) {
					const temp: number = start;
					start = end;
					end = temp;
				}
				assertEquals(cums[end] - cums[start], bt.getRangeSum(start, end));
			}
		}
	},
	
];



/*---- Helper definitions ----*/

function randInt(n: number): number {
	return Math.floor(Math.random() * n);
}


function assertEquals(expect: number, actual: number): void {
	if (actual !== expect)
		throw new Error("Value mismatch");
}



/*---- Main runner ----*/

(function(): void {
	let i: number = 0;
	function iterate(): void {
		let msg: string;
		if (i >= TEST_SUITE_FUNCS.length)
			msg = "Finished";
		else {
			msg = TEST_SUITE_FUNCS[i].name + "(): ";
			try {
				TEST_SUITE_FUNCS[i]();
				msg += "Pass";
			} catch (e) {
				msg += "Fail - " + e.message;
			}
			i++;
			setTimeout(iterate);
		}
		let li: HTMLElement = document.createElement("li");
		li.textContent = msg;
		(document.getElementById("results") as HTMLElement).append(li);
	}
	iterate();
})();
